In [1]:
# https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud?select=creditcard.csv

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression

In [2]:
data = pd.read_csv("F:\MachineLearningProjects\creditcard.csv") # File Path

In [6]:
data.head() #princple component analysis

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
data.shape

(284807, 31)

In [7]:
data.isnull().sum().sum()

0

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [9]:
# here it comes unbalanced 0-->normal, 1-->fraud
data.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [10]:
legit = data[data.Class == 0]
fraud = data[data.Class==1]

In [11]:
legit.Class.value_counts()

0    284315
Name: Class, dtype: int64

In [12]:
fraud.Class.value_counts()

1    492
Name: Class, dtype: int64

In [13]:
legit.shape,fraud.shape

((284315, 31), (492, 31))

In [14]:
#statistial measures
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [15]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [18]:
# Compare values of legit and fraud data
data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

### Under Sampling

Build a sample dataset containing similar distrubution of noraml trans and fraud trans

fraud = 492<br>
legit = 284315

In [19]:
legit_sample= legit.sample(n=492)


In [20]:
#concatinate fraud data and legit data
concat_data = pd.concat([legit_sample,fraud],axis=0)

In [21]:
concat_data.head()

Time        V1        V2        V3        V4        V5        V6  \
28515    35043.0 -0.793405  0.491014  2.126645 -0.028050  0.277626  1.113498   
43330    41464.0 -4.001754  0.974763  0.133843  0.046463 -0.226866 -1.389429   
172097  120958.0 -0.368529  1.199197 -0.283921 -0.842756  1.441302 -0.557406   
69606    53500.0  0.927986 -0.557070  1.364354  1.425934 -0.943806  0.963711   
217295  140842.0 -1.378732 -0.619474 -2.778610 -1.354187 -4.106769  3.483853   

              V7        V8        V9  ...       V21       V22       V23  \
28515   0.022028  0.567925  0.631236  ... -0.104737 -0.297590 -0.407138   
43330   0.275690  0.152493  1.194751  ... -0.873139 -0.529152  1.017438   
172097  1.536815 -0.552978 -0.045209  ...  0.188872  0.772217 -0.567976   
69606  -0.919989  0.370814  1.209447  ...  0.226804  0.761376 -0.233443   
217295  4.133103  0.283119 -1.836258  ...  0.064822  0.673974  0.945652   

             V24       V25       V26       V27       V28   Amount  Class  
28515  -1.402547  0.447647 -0.352422  0.082768  0.056397    12.00      0  
43330   0.360270  0.387922  0.105846  0.437283 -0.198167    10.74      0  
172097 -1.121446  0.231127 -0.121674 -0.265836 -0.066922    30.90      0  
69606  -0.409133  0.434014 -0.122192  0.096001  0.051253   105.50      0  
217295 -0.516301 -0.582480  0.753626  0.159631 -0.304107  1093.99      0  

[5 rows x 31 columns]

In [23]:
concat_data['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [24]:
# Splitting
X = concat_data.drop('Class',axis=1)

In [25]:
y = concat_data.Class

In [26]:
X.shape,y.shape

((984, 30), (984,))

In [27]:
# train test split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=.2,stratify=y,random_state=1)

In [28]:
model = LogisticRegression()

In [29]:
model.fit(X_train,y_train)

C:\Users\hi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [30]:
test_pred = model.predict(X_test)

In [31]:
train_pred = model.predict(X_train)

In [33]:
accuracy_score(y_test,test_pred)

0.9238578680203046

In [34]:
accuracy_score(y_train,train_pred)

0.9466327827191868

## testing model

In [46]:
inp = (472,1.01705496852431,-0.185048783070119,1.18166606650006,1.07349020426383,-0.550586426062338,0.971826768161323,-0.80645193091673,0.4847922444553,0.517268097462824,-0.06012395521202,1.44699044339251,1.19801345982087,-0.0708907835143611,0.00657736573307565,0.432215270517664,-0.14987287415034,-0.0908413936683139,-0.19527724457351,-0.870040663807092,-0.164292863021539,0.239286567271276,0.875847990785814,-0.0534575789685018,-0.269918491299512,0.328906924068146,-0.217332071615722,0.0962076194960413,0.0207962233490798,17.57)

In [47]:
len(inp)

30

In [48]:
inp_np = np.asarray(inp)

In [49]:
inputdata = inp_np.reshape(1,-1)

In [50]:
model.predict(inputdata)

C:\Users\hi\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


array([0], dtype=int64)